Add extra options for show/hide all sections accordion click

patrickpatrickpatrick opened the pull request:
## What

Adds ability to add extra options to the click anayltics event for showing/hiding all the sections of an accordion.

## Why

In a [previous PR ](https://github.com/alphagov/govuk_publishing_components/pull/2813)I added the ability to add extra options to click events for opening sections in an accordion but I did not include the Show/Hide All Sections button. For tracking purposes it means that if an attribute is not passed in an event, then an event cannot be filtered on that attribute. Therefore adding the ability to pass attributes to the Show/Hide All Sections button means it can be more easily inlcuded in searches/filters on click events.

[Relevant Trello Card](https://trello.com/c/hDJRk2hS/1096-add-tracking-for-position-of-show-hide-all-sections-in-accordion-on-new-browse-level-2-curated-topics)

andysellick commented on app/views/govuk_publishing_components/components/docs/accordion.yml:
```diff
@@ -189,7 +189,7 @@ examples:
       Each item can also have a `data_attributes` hash. These `data_attributes` are placed on the `button` that triggers the opening and closing - useful for differentiating between each section of the accordion.
 
-      Data attributes can also be added to the 'Show/hide all' link using the `data_attributes_show_all` option, primarily where custom tracking is required. These attributes are read from the accordion markup and then added to the link by JavaScript (which is how the link is created).
+      Data attributes can also be added to the 'Show/hide all' link using the `data_attributes_show_all` option, primarily where custom tracking is required. These attributes are read from the accordion markup and then added to the link by JavaScript (which is how the link is created). If `track_options` within `data_attributes_show_all` is set, then it is possible to pass a custom dimension when ...
```
Can you add an example of this to the documentation? (you might need to create a separate entry)

andysellick commented on spec/javascripts/components/accordion-spec.js:
```diff
@@ -49,22 +63,50 @@ describe('Accordion component', function () {
+      }
+    }
+
+    accordion.setAttribute('data-show-all-attributes', JSON.stringify(object))
```
Could this technique (of setting the required attributes on the fixture in the test itself, rather than the markup at the top of the file) be used in the previous test as well? It makes the test more self-contained and readable.

andysellick commented on spec/javascripts/components/accordion-spec.js:
```diff
@@ -49,22 +63,50 @@ describe('Accordion component', function () {
     var wrappingObject = {
       'gtm-event-name': 'example',
-      'show-all-attributes': JSON.stringify(object)
+      ...object
```
Is there a particular reason for using spread syntax here, over the previous method? I ask because it's not supported in IE and it would be good to be able to run the tests in IE if we need to, and this would block that.

andysellick commented on spec/javascripts/components/accordion-spec.js:
```diff
@@ -49,22 +63,50 @@ describe('Accordion component', function () {
+    expect(document.querySelector('.govuk-accordion__show-all').getAttribute('data-track-options')).toEqual(JSON.stringify({ dimensionExample: 'Example value' }))
     expect(document.querySelector('.govuk-accordion__show-all').getAttribute('data-gtm-event-name')).toEqual('example')
-    expect(document.querySelector('.govuk-accordion__show-all').getAttribute('data-show-all-attributes')).toEqual(JSON.stringify(object))
+    expect(document.querySelector('.govuk-accordion__show-all').getAttribute('data-ui')).toEqual(JSON.stringify(object.ui))
```
I think this is modifying the intent of the original test, which is to test that the show/hide all link has been correctly given a `data-show-all-attributes` attribute (needed for the GA4 work).

Suggestion - could this test be reverted and a new test added to check this new functionality? It feels related but distinct.

andysellick commented on spec/javascripts/components/accordion-spec.js:
```diff
@@ -7,40 +7,54 @@ describe('Accordion component', function () {
 
   beforeEach(function () {
+    spyOn(GOVUK.analytics, 'trackEvent')
+    spyOn(GOVUK.Modules, 'GemTrackClick').and.callFake(function () { this.init = function () {} })
```
I can't see where `GemTrackClick` is referenced - can you explain more about why this line is needed?

andysellick commented on spec/javascripts/components/accordion-spec.js:
```diff
@@ -7,40 +7,54 @@ describe('Accordion component', function () {
+                  <span class="govuk-accordion__section-toggle-focus">
+                    <span class="govuk-accordion-nav__chevron govuk-accordion-nav__chevron--down"></span>
+                    <span class="govuk-accordion__section-toggle-text">Show<span class="govuk-visually-hidden"> this section</span>
+                  </span>
```
It looks like these additional lines are to make the markup closer to the actual rendered content, but is there a particular reason for this? I'd generally recommend that fixture markup is as simple as possible - if these extra spans are needed for a test then fair enough, but I can't see them being referenced anywhere.

andysellick reviewed (commented):
Thanks for adding some tests, although I've raised a few questions. Also would it be possible to add some tests for https://github.com/alphagov/govuk_publishing_components/pull/2813? (either in this PR as a separate commit or in a separate PR, whichever makes sense).

patrickpatrickpatrick commented:
Moved all new tests to separate tests and deleted previous changes to existing ones. Reverted all ES6 changes and removed extraneous mark-up and statements.

> Thanks for adding some tests, although I've raised a few questions. Also would it be possible to add some tests for #2813? (either in this PR as a separate commit or in a separate PR, whichever makes sense).

`it('applies custom attributes to click event if specified in section')` should test the change I made in that PR, as the additional data attribute for the sections are added directly to the element and aren't appended using javascript like for the show all sections data attributes

andysellick commented on app/assets/javascripts/govuk_publishing_components/components/accordion.js:
```diff
@@ -48,13 +48,18 @@ window.GOVUK.Modules.GovukAccordion = window.GOVUKFrontend.Accordion;
+            showAll.setAttribute('data-' + keys[i], JSON.stringify(values[keys[i]]))
+          } else {
+            showAll.setAttribute('data-' + keys[i], values[keys[i]])
+          }
```
As discussed, turns out this change isn't necessary - the existing code will already append any given attribute onto the show all element.

andysellick reviewed (commented):
Almost ready to approve! Just one thing, which we just discussed.

andysellick reviewed (approved):
🚀
